packages/eslint-plugin/src/rules/no-use-before-define.ts

```diff
@@ -92,6 +92,29 @@ function isOuterVariable(
   );
 }
 
+/**
+ * Recursively checks whether or not a given reference has a type query declaration among it's parents
+ */
+function referenceContainsTypeQuery(node: TSESTree.Node): boolean {
+  if (node.type === AST_NODE_TYPES.TSTypeQuery) {
+    return true;
+  } else if (!node.parent) {
+    return false;
+  }
+
+  return referenceContainsTypeQuery(node.parent);
```

Comment:
Almost there!

We can constrain this a little more so we don't recursively traverse up the whole tree unnecessarily.
[We know that the node must always be either be a `TSQualifiedName`, `Identifier` or `TSTypeQuery`](https://github.com/typescript-eslint/typescript-eslint/pull/2623#discussion_r498631601).

```ts
switch (node.type) {
  case AST_NODE_TYPES.TSTypeQuery:
    return true;

  case AST_NODE_TYPES.TSQualifiedName:
  case AST_NODE_TYPES.Identifier:
    return referenceContainsTypeQuery(node.parent);

  default:
    // if we find a different node, there's no chance that we're in a TSTypeQuery
    return false;
}
```

Reply:
I see! I wasn't sure of what you meant in your first comment.
I've changed it.